field/TinyLang/Field/UParser.hs

```diff
@@ -0,0 +1,258 @@
+prefixed by @#@.  We will check if an identifier is not a keyword.
+
+@
+ident ::= ( "?" | "#" ) [a-z] ([a-z0-9_'])*
+@
+
+== Constants
+
+At the moment constants consist of literals only.
+
+@
+const ::=
+  bool-literal
+  int-literal
+  vec-literal
+@
+
+== Expressions
+
+As mentioned above @expr@ is the designated start production for the
+parser.
+
+@
+expr ::=
+  const
+  "(" expr ")"
+  ident
+  expr infix-op expr
+  prefix-op expr
+  expr "[" expr "]"
+  statement ";" expr
+  "if" expr "then" expr "else" expr
+
+infix-op ::=
+  "and"
+  "or"
+  "xor"
+  "=="
+  "<="
+  "<"
+  ">="
+  ">"
+  "+"
+  "-"
+  "*"
+  "/"
+
+prefix-op ::=
+  "not"
+  "neq0"
+  "neg"
+  "inv"
+  "unpack"
+@
+
+== Statement
+
+Statements are a bit odd at the moment, as they are neither
+expressions nor the usual statements.
+
+@
+statement ::=
+  "let" var "=" expr
+  "assert" expr
+  "for" var "=" int "to" int "do" statements "end"
+
+statements ::=
+  (statement (";" statement)*)?
+@
+
+== Operator Precedence
+  Precedence: "not" > "xor" > "and" > "or"  (but use parentheses anyway).
+  if-then-else has to be parenthesised unless it's at the very top.
+
+  Precedence for numeric operators is standard:  {neg,inv} > {*,/} > {+,- }.
+  Things like "neg inv 5" are illegal: use parentheses.
+
+  The code is based on the tutorial at
+  https://markkarpov.com/megaparsec/parsing-simple-imperative-language.html
+
+  See also https://markkarpov.com/megaparsec/megaparsec.html
+-}
+
+
+module TinyLang.Field.UParser where
+
+import           Control.Applicative.Combinators hiding (many, sepBy, some)
+import           Text.Megaparsec
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer      as L
+import           Data.Void
+import qualified Data.Set                        as S
+import           GHC.Unicode ( isLower
+                             , isDigit)
+
+type Parser = Parsec Void String
+
+type Identifier = String
+
+data Const
+  = CBool  Bool
+  | CInt   Integer
+  | CVec   [Bool]
+
+data Var = Var Identifier
+
+data Expr
+  = EConst     Const
+  | EVar       Var
+  | EAppBinOp  UnOp      Expr
+  | EAppUnOp   BinOp     Expr
+  | EStatement Statement Expr
+  | EIf        Expr      Expr Expr
+
+data BinOp
+  = Or
+  | And
+  | Xor
+  | FEq
+  | FLe
+  | FGe
+  | FGt
+  | Add
+  | Sub
+  | Mul
+  | Div
+  | BAt
+
+data UnOp
+  = Not
+  | Neq0
+  | Neg
+  | Inv
+  | Unp
+  
+data Statement
+  = ELet    Var Expr
+  | EAssert Expr
+
+-- Lexer
+sc :: Parser ()
+sc = L.space space1 empty empty
+
+lexeme :: Parser a -> Parser a
+lexeme = L.lexeme sc
+
+symbol :: String -> Parser String
+symbol = L.symbol sc
+
+
+-- Identifier Character
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c =
+  or $ map (\f -> f c) [isLower , isDigit, (=='_'), (=='\'')]
+
+
+identifierChar :: (MonadParsec e s m, Token s ~ Char) => m (Token s)
+identifierChar = satisfy isIdentifierChar <?> "identifier character [a-z0-9_']"
+{-# INLINE identifierChar #-}
```

Comment:
Adding `INLINE` pragmas should be done together with benchmarking (I've seen cases where my code became slower after I sprinkled `INLINE` pragmas over it where I thought they would make sense), so don't bother adding this syntactic noise right now.

Reply:
:+1: